datafusion/physical-expr/src/string_map.rs

```diff
@@ -0,0 +1,1028 @@
+                else {
+                    // Put the small values into buffer and offsets so it appears
+                    // the output array, but store the actual bytes inline for
+                    // comparison
+                    self.buffer.append_slice(value);
+                    self.offsets.push(O::from_usize(self.buffer.len()).unwrap());
+                    let payload = make_payload_fn(Some(value));
+                    let new_header = Entry {
+                        hash,
+                        len: value_len,
+                        offset_or_inline: inline,
+                        payload,
+                    };
+                    self.map.insert_accounted(
+                        new_header,
+                        |header| header.hash,
+                        &mut self.map_size,
+                    );
+                    payload
+                }
+            }
+            // value is not a "small" string
+            else {
+                // Check if the value is already present in the set
+                let entry = self.map.get_mut(hash, |header| {
+                    // compare value if hashes match
+                    if header.len != value_len {
+                        return false;
+                    }
+                    // Need to compare the bytes in the buffer
+                    // SAFETY: buffer is only appended to, and we correctly inserted values and offsets
+                    let existing_value =
+                        unsafe { self.buffer.as_slice().get_unchecked(header.range()) };
+                    value == existing_value
+                });
+
+                if let Some(entry) = entry {
+                    entry.payload
+                }
+                // if no existing entry, make a new one
+                else {
+                    // Put the small values into buffer and offsets so it
+                    // appears the output array, and store that offset
+                    // so the bytes can be compared if needed
+                    let offset = self.buffer.len(); // offset of start fof data
+                    self.buffer.append_slice(value);
+                    self.offsets.push(O::from_usize(self.buffer.len()).unwrap());
+
+                    let payload = make_payload_fn(Some(value));
+                    let new_header = Entry {
+                        hash,
+                        len: value_len,
+                        offset_or_inline: offset,
+                        payload,
+                    };
+                    self.map.insert_accounted(
+                        new_header,
+                        |header| header.hash,
+                        &mut self.map_size,
+                    );
+                    payload
+                }
+            };
+            observe_payload_fn(payload);
+        }
+    }
+
+    /// Removes the first n distinct values inserted into this set, in the order
```

Comment:
This is the code that removes only the first_n from the set and updates the state. It is pretty complicated and I am not sure the performance benefit is worth it

Reply:
It does have significant test coverage, however